### PR TITLE
UHV: Move checks for headers with underscores to UHV when it is enabled

### PR DESCRIPTION
**Pull request by yanavlasov**
Commit Message:
While UHV has implemented checks for header names with underscores, the corresponding checks in codec_impl.cc was not disabled. This PR removes this check from codecs when ENVOY_ENABLE_UHV is defined, such that code in UHV is exercised.
Also added missing plumbing for the stats for header map modification done by UHV.

Risk Level: Low, main build is not affected
Testing: Unit Tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

**Review by alyssawilk — DISMISSED** [review]
this feels like a lot of complexity for a solution aimed to simplify but I can't think of any improvements :-/
please do get !google sign-off as well.

**Review by lizan — APPROVED** [review]
API LGTM and non-Googler sign-off.

File: envoy/http/header_validator.h
Comment (alyssawilk) [inline envoy/http/header_validator.h:109]: optional: would it make more sense to leave these pure, remove the guard around the implementation, and add guards around invocation?
If you want to leave as-is I think a TODO to remove base class impl a part of the UHV bug would be good.

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo outstanding nit.  I'll let Kevin make sure that's sorted one way or another for final stamp before merging since I'm out soon.

File: envoy/http/header_validator.h
Comment (yanavlasov) [reply envoy/http/header_validator.h]: I agree. I realized I could fix coverage by using these methods in the ifdefed code.

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by RyanTheOptimist — APPROVED** [review]
Coverage changes LGTM

**Pull request merged**
